### PR TITLE
test: add provisioner and nodeclaim counter suite tests

### LINKED ISSUE #489: Test Coverage for the Provisioner counter controller
### Description
https://github.com/aws/karpenter-core/tree/main/pkg/controllers/provisioner/counter doesn't have a suite_test.go file, which means there aren't any suite_tests for this. This status is used to know how close a user's Provisioner provisioned capacity is to its limits.

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by njtran**
Fixes #489 

**Description**
Adds in suite tests for the counter controller.

**How was this change tested?**
- make presubmit

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/test/expectations/expectations.go [hunk 1]
```diff
@@ -599,3 +600,33 @@ func ExpectMachinesWithOffset(offset int, ctx context.Context, c client.Client)
 		return &m
 	})
 }
+
+func ExpectMakeNodesMachinesInitializedAndStateUpdated(ctx context.Context, c client.Client, nodeStateController, machineStateController controller.Controller, nodes []*v1.Node, machines []*v1alpha5.Machine) {
```
Comment (jonathan-innis) [inline pkg/test/expectations/expectations.go:604]: Should we just break up these functions to be? 

`ExpectNodesInitializedAndStateUpdated`
`ExpectMachinesInitializedAndStateUpdated`
`ExpectNodeClaimsInitializedAndStateUpdated`

I'm wondering if they're getting far too wordy and it might be best to just decouple them

File: pkg/controllers/nodepool/counter/suite_test.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+var machineController controller.Controller
+var nodeController controller.Controller
+var podController controller.Controller
+var ctx context.Context
+var env *test.Environment
+var cluster *state.Cluster
+var fakeClock *clock.FakeClock
+var cloudProvider *fake.CloudProvider
+
+var provisioner *v1alpha5.Provisioner
```
Comment (jonathan-innis) [inline pkg/controllers/nodepool/counter/suite_test.go:56]: Should test be globals or scoped into their `Describe` blocks?

File: pkg/controllers/nodepool/counter/provisioner_test.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+				},
+			},
+		})
+		ExpectApplied(ctx, env.Client, provisioner)
+		ExpectReconcileSucceeded(ctx, provisionerInformerController, client.ObjectKeyFromObject(provisioner))
+		ExpectReconcileSucceeded(ctx, provisionerController, client.ObjectKeyFromObject(provisioner))
+		provisioner = ExpectExists(ctx, env.Client, provisioner)
+	})
+
+	It("should increase the counter when a new nodes are created", func() {
```
Comment (jonathan-innis) [inline pkg/controllers/nodepool/counter/provisioner_test.go:70]:
```suggestion
	It("should increase the counter when new nodes are created", func() {
```

File: pkg/controllers/nodepool/counter/provisioner_test.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		ExpectApplied(ctx, env.Client, node, machine)
+		ExpectMakeNodesMachinesInitializedAndStateUpdated(ctx, env.Client, nodeController, machineController, []*v1.Node{node}, []*v1alpha5.Machine{machine})
+
+		ExpectReconcileSucceeded(ctx, provisionerController, client.ObjectKeyFromObject(provisioner))
+		provisioner = ExpectExists(ctx, env.Client, provisioner)
+
+		// Should equal both the machine and node capacity
+		Expect(provisioner.Status.Resources).To(BeEquivalentTo(machine.Status.Capacity))
```
Comment (jonathan-innis) [inline pkg/controllers/nodepool/counter/provisioner_test.go:78]: I think generally speaking, the Machine and Node capacity won't match. Maybe to make this more realistic, we should:
1. Add a machine
2. Check the provisioner status
3. Register a node and initialize it
4. Check the provisioner status to ensure it now matches the node

File: pkg/controllers/nodepool/counter/provisioner_test.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+		Expect(provisioner.Status.Resources).To(BeEquivalentTo(node.Status.Capacity))
+	})
+	It("should decrease the counter when an existing node is deleted", func() {
```
Comment (jonathan-innis) [inline pkg/controllers/nodepool/counter/provisioner_test.go:81]: Nice 🎉

File: pkg/test/expectations/expectations.go (hunk 1 above)
Comment (njtran) [reply pkg/test/expectations/expectations.go:604]: Yeah, personally, I think this is acceptable, and we merged these functions together since the deprovisioning code calls the ExpectNodes... and ExpectMachines... call a lot. 

I renamed it to be ...NodeNodeClaims... because of the fact that we have to replicate the Provisioner/NodePool Machine/NodeClaim logic across the codebase. When we remove the instances of Provisioners/Machines from the codebase, I think we can rename this back to ExpectMakeInitializedAndStateUpdated(). Thoughts?

File: pkg/controllers/nodepool/counter/provisioner_test.go (hunk 4 above)
Comment (njtran) [reply pkg/controllers/nodepool/counter/provisioner_test.go:114]: Good callout! Added another test here.

File: pkg/test/expectations/expectations.go (hunk 1 above)
Comment (jonathan-innis) [reply pkg/test/expectations/expectations.go:604]: Sure, I might opt for throwing the word `And` in here so that it's a little more readable, but I'm fine whatever name you decide:

```go
ExpectMakeNodesAndMachinesInitializedAndStateUpdated
ExpectMakeNodeClaimsAndMachineInitializedAndStateUpdated
```

File: pkg/controllers/nodepool/counter/provisioner_test.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+		// Don't initialize the node yet
+		ExpectMakeMachinesInitialized(ctx, env.Client, machine)
+		// Inform cluster state about node and machine readiness
+		ExpectReconcileSucceeded(ctx, nodeController, client.ObjectKeyFromObject(node))
+		ExpectReconcileSucceeded(ctx, machineController, client.ObjectKeyFromObject(machine))
+
+		ExpectReconcileSucceeded(ctx, provisionerController, client.ObjectKeyFromObject(provisioner))
+		provisioner = ExpectExists(ctx, env.Client, provisioner)
+
+		// Should equal both the machine and node capacity
```
Comment (jonathan-innis) [inline pkg/controllers/nodepool/counter/provisioner_test.go:84]: I think this comment might be a little out of date

**Review by jonathan-innis — DISMISSED** [review]
LGTM, only a couple nits

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀
